test/boilerplate_suggestors/advanced_props_and_state_class_migrator_test.dart

```diff
@@ -0,0 +1,455 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:over_react_codemod/src/boilerplate_suggestors/advanced_props_and_state_class_migrator.dart';
+import 'package:over_react_codemod/src/boilerplate_suggestors/boilerplate_utilities.dart';
+import 'package:test/test.dart';
+
+import '../util.dart';
+
+void main() {
+  group('AdvancedPropsAndStateClassMigrator', () {
+    final testSuggestor =
+        getSuggestorTester(AdvancedPropsAndStateClassMigrator());
+
+    tearDown(() {
+      propsAndStateClassNamesConvertedToNewBoilerplate = {};
+    });
+
+    group('does not operate when', () {
+      test('it\'s an empty file', () {
+        testSuggestor(expectedPatchCount: 0, input: '');
+
+        expect(propsAndStateClassNamesConvertedToNewBoilerplate, isEmpty);
+      });
+
+      test('the class is simple', () {
+        testSuggestor(
+          expectedPatchCount: 0,
+          input: r'''
+        @Factory()
+        UiFactory<FooProps> Foo =
+            // ignore: undefined_identifier
+            $Foo;
+
+        @Props()
+        class _$FooProps extends UiProps {
+          String foo;
+          int bar;
+        }
+
+        @Component2()
+        class FooComponent extends UiComponent2<FooProps> {
+          @override
+          render() {
+            return Dom.ul()(
+              Dom.li()('Foo: ', props.foo),
+              Dom.li()('Bar: ', props.bar),
+            );
+          }
+        }
+      ''',
+        );
+
+        expect(propsAndStateClassNamesConvertedToNewBoilerplate, isEmpty);
+      });
+
+      test(
+          'the class is advanced and the classes extend from a custom class in addition to having mixins',
+          () {
+        testSuggestor(
+          expectedPatchCount: 0,
+          input: r'''
+      @Factory()
+      UiFactory<FooProps> Foo =
+          // ignore: undefined_identifier
+          $Foo;
+
+      @Props()
+      class _$FooProps extends ADifferentPropsClass with AMixin, AnotherMixin {
```

Comment:
What about the case where `ADifferentPropsClass` is the one with mixins?

Reply:
Offline resolution - always add a `// FIXME` comment when a class extends from a class that is not `UiProps`.